nengo/builder/optimizer.py

```diff
@@ -0,0 +1,911 @@
+        # performed on at most one type of operator per pass.
+        # The depedency graph and other information will be updated
+        # before merging other operator types.
+
+        # We go through ops in a heuristic order to reduce runtime
+        firstops = [ElementwiseInc, SlicedCopy, DotInc, SimNeurons]
+        sortedops = firstops + [op for op in by_type if op not in firstops]
+        for optype in sortedops:
+
+            if OpMerger.is_type_mergeable(optype):
+                self.perform_merges_for_subset(by_type[optype])
+
+            # If we're not only merging views, the memory layout changes
+            # and non-views are turned into views. In that case we need
+            # to update the signals the operators are referring to before
+            # trying to merge a different type of operators. Thus, we break
+            # the loop here.
+            if not self.only_merge_ops_with_view and len(self.merged) > 0:
+                break
+
+        # At this point, we may have marked some ops for merge.
+        # Some signals may therefore have been replaced.
+        # Those signals might be used in some other op, so we go through
+        # all other ops and do the same signal replacement if necessary.
+        self.resolve_views_on_replaced_signals()
+
+    def perform_merges_for_subset(self, subset):
+        """Performs operator merges for a subset of operators.
+
+        Parameters
+        ----------
+        subset : list
+            Subset of operators.
+        """
+        by_view = groupby(subset, lambda op: self.opinfo[op].v_base)
+        if self.only_merge_ops_with_view and None in by_view:
+            # If an op has no views, v_base will be None.
+            # If we're only merging views, then we get rid of this subset.
+            del by_view[None]
+
+        for view_subset in itervalues(by_view):
+            if len(view_subset) > 1:
+                self.perform_merges_for_view_subset(view_subset)
+
+    def perform_merges_for_view_subset(self, subset):
+        """Perform merges for a subset of operators with the same view base.
+
+        Parameters
+        ----------
+        subset : list
+            Subset of operators. These need to have the same view base (can be
+            None if it is None for all) for their first signal in
+            `all_signals`.
+        """
+
+        # Sort to have sequential memory.
+        offsets = np.array([self.opinfo[op].v_offset for op in subset])
+        sort_indices = np.argsort(offsets)
+        offsets = offsets[sort_indices]
+        sorted_subset = [subset[i] for i in sort_indices]
+
+        for i, op1 in enumerate(sorted_subset):
+            if op1 in self.merged:
+                # Cannot merge merged operator again until dependency graph
+                # has been updated
+                continue
+
+            if any(op1 in self.dependents[op] or op in self.dependents[op1]
+                   for op in self.merged):
```

Comment:
I think this loop is the (or one of the) reasons for the performance problem with Spaun. I will try if it improves things to keep around a `set` with the union of all `self.dependents[op]` and switch the second part to `op in self.merged for op in self.dependents[op1]`.